[llvm][Support] Improvements to ListeningSocket functionality and documentation

cpsughrue opened the pull request:
Improvements include
* Enable `ListeningSocket::accept` to timeout after a specified amount of time or block indefinitely
* Enable `ListeningSocket::createUnix` to handle instances where the target socket address already exists and differentiate between situations where the existing file does and does not already have a bound socket
* Doxygen comments

Functionality added for the module build daemon

cpsughrue committed:
[llvm][Support] Improvements to raw_socket_stream functionality and documentation

cpsughrue committed:
Only close PipeFDs when they do not equal -1

cpsughrue committed:
Use atomic operations on file descriptor check

Co-authored-by: Michael Spencer <bigcheesegs@gmail.com>

cpsughrue committed:
Rename two functions, shorten timeout during test, and fix comment typo

cpsughrue committed:
Handle case were poll exists due to signal

Bigcheese commented on llvm/lib/Support/raw_socket_stream.cpp:
```diff
@@ -56,104 +61,222 @@ static std::error_code getLastSocketErrorCode() {
+    return;
+  if (!FD.compare_exchange_strong(ObservedFD, -1))
     return;
   ::close(FD);
```
This needs to close `ObservedFD`. `FD` will always be -1 here.

Bigcheese commented on llvm/lib/Support/raw_socket_stream.cpp:
```diff
@@ -56,104 +61,222 @@ static std::error_code getLastSocketErrorCode() {
+      return llvm::make_error<StringError>(
+          std::make_error_code(std::errc::timed_out),
+          "No client requests within timeout window");
+
```
You need to check if the pipe was written to here:
```
if (FDs[1].revents & POLLIN)
  return llvm::make_error<StringError>(
          std::make_error_code(std::errc::operation_canceled),
          "Accept canceled");
```

Bigcheese commented on llvm/unittests/Support/raw_socket_stream_test.cpp:
```diff
@@ -58,11 +59,80 @@ TEST(raw_socket_streamTest, CLIENT_TO_SERVER_AND_SERVER_TO_CLIENT) {
+  // Create a separate thread to close the socket after a delay. Simulates a
+  // signal handler calling ServerListener::shutdown
+  std::thread CloseThread([&]() {
+    std::this_thread::sleep_for(std::chrono::seconds(2));
```
Use a shorter time to reduce test run length.

Bigcheese commented on llvm/unittests/Support/raw_socket_stream_test.cpp:
```diff
@@ -58,11 +59,80 @@ TEST(raw_socket_streamTest, CLIENT_TO_SERVER_AND_SERVER_TO_CLIENT) {
+  llvm::Error Err = MaybeServer.takeError();
+  llvm::handleAllErrors(std::move(Err), [&](const llvm::StringError &SE) {
+    std::error_code EC = SE.convertToErrorCode();
+    ASSERT_EQ(EC, std::errc::bad_file_descriptor);
```
With the change to detect the pipe write this will instead see `operation_canceled`.

cpsughrue committed:
Address feedback

Bigcheese reviewed (approved):
Looks good with the spelling fix. Thanks!

cpsughrue committed:
Fix type and remove \brief from doxygen comments

cpsughrue committed:
Remove \class from doxygen comments and correct PipeFD comment

cpsughrue committed:
Remove ASSERT that checks the duration of function call

Occasionally the unit test, TIMEOUT_PROVIDED, would fail because the
allowed margin of error when comparing Timeout with Duration was not
large enough. Instead of trying to guess an appropriate error margin
and risk creating a flaky test solely rely on checking the returned
error code to confirm that ListeningSocket::accept timed out

cpsughrue commented:
![image](https://github.com/llvm/llvm-project/assets/55301806/63666c05-e1f2-4edf-b982-8359e620e28a)
During pre-commit CI the unit test,`TIMEOUT_PROVIDED`, failed because the allowed margin of error when comparing `Duration` to `Timeout` was not large enough. @Bigcheese, what do you think about removing the ASSERT that compares `Duration` to `Timeout` and instead solely relying on checking that the returned error code is `std::errc:timed_out`?

Bigcheese commented:
I think that's fine. In general for timeout tests I would just check that the duration is >= the timeout, and rely on manual testing to make sure it's not too much longer. It's not trivial to figure out a reasonable maximum time for a timeout to take in these kinds of cases, and here knowing that the OS says it timed out is mostly what we care about.

cpsughrue commented:
Sounds good

rupprecht commented on llvm/lib/Support/raw_socket_stream.cpp:
```diff
@@ -56,104 +61,231 @@ static std::error_code getLastSocketErrorCode() {
-#endif // _WIN32
+  // Ensure ::poll returns if shutdown is called by a seperate thread
+  char Byte = 'A';
+  ::write(PipeFD[1], &Byte, 1);
```
This causes `-Wunused-result`, suppressed by 6b46166ef2612d2a58767447b3db8f0343afb552.

Is there any meaningful way to recover from an error returned by write? Or is it even relevant at all whether it fails? If no recovery is necessary, 6b46166ef2612d2a58767447b3db8f0343afb552 should be sufficient. If you'd like to propagate some error or log something in case of failure, you may want to follow up.

cpsughrue replied:
6b46166 should be enough. Thanks for the fix!
